src/ddmd/link.d

```diff
@@ -981,62 +925,409 @@ public int runProgram()
+            {
+                const(char)* arch = x64 ? "x64" : "x86";
+                auto sdk = FileName.combine(WindowsSdkDir, "lib");
+                if (FileName.exists(FileName.buildPath(sdk, WindowsSdkVersion, "um", arch, "kernel32.lib"))) // SDK 10.0
+                    return FileName.buildPath(sdk, WindowsSdkVersion, "um", arch);
+                else if (FileName.exists(FileName.buildPath(sdk, r"win8\um", arch, "kernel32.lib"))) // SDK 8.0
+                    return FileName.buildPath(sdk, r"win8\um", arch);
+                else if (FileName.exists(FileName.buildPath(sdk, r"winv6.3\um", arch, "kernel32.lib"))) // SDK 8.1
+                    return FileName.buildPath(sdk, r"winv6.3\um", arch);
+                else if (x64 && FileName.exists(FileName.buildPath(sdk, arch, "kernel32.lib"))) // SDK 7.1 or earlier
+                    return FileName.buildPath(sdk, arch);
+                else if (!x64 && FileName.exists(FileName.buildPath(sdk, "kernel32.lib"))) // SDK 7.1 or earlier
+                    return sdk;
+            }
+            return null;
+        }
+
+        // iterate through subdirectories named by SDK version in baseDir and return the
+        //  one with the largest version that also contains the test file
+        static const(char)* findLatestSDKDir(const(char)* baseDir, const(char)* testfile)
+        {
+            auto allfiles = FileName.combine(baseDir, "*");
+            static if (!is(WIN32_FIND_DATAA)) alias WIN32_FIND_DATAA = WIN32_FIND_DATA; // support dmd 2.068
+            WIN32_FIND_DATAA fileinfo;
+            HANDLE h = FindFirstFileA(allfiles, &fileinfo);
+            if (h == INVALID_HANDLE_VALUE)
+                return null;
+
+            char* res = null;
+            do
+            {
+                if (fileinfo.cFileName[0] >= '1' && fileinfo.cFileName[0] <= '9')
+                    if (res is null || strcmp(res, fileinfo.cFileName.ptr) < 0)
+                        if (FileName.exists(FileName.buildPath(baseDir, fileinfo.cFileName.ptr, testfile)))
+                        {
+                            const len = strlen(fileinfo.cFileName.ptr) + 1;
+                            res = cast(char*) memcpy(mem.xrealloc(res, len), fileinfo.cFileName.ptr, len);
+                        }
+            }
+            while(FindNextFileA(h, &fileinfo));
+            FindClose(h);
+
+            return res;
+        }
+
+        pragma(lib, "advapi32.lib");
+
+        /**
+         * read a string from the 32-bit registry
+         * Params:
+         *  softwareKeyPath = path below HKLM\SOFTWARE
+         *  valueName       = name of the value to read
+         * Returns:
+         *  the registry value if it exists and has string type
+         */
+        const(char)* GetRegistryString(const(char)* softwareKeyPath, const(char)* valueName)
+        {
+            enum x64hive = false; // VS registry entries always in 32-bit hive
+
+            version(Win64)
```

Comment:
Wait a minute! Why would the 64 bit build of DMD behave differently here? Shouldn't this be if the memory model of the target is set to 64?

Reply:
No, the target architecture is not relevant here. The VS registry/Windows SDK keys are only found in the 32-bit registry hive. KEY_WOW64_32KEY doesn't seem good enough, I had to use WOW6432Node explicitly, while a 32-bit process is automatically mapped to that hive.